### PR TITLE
fix request for cpp exceptions

### PR DESCRIPTION
**Pull request by ihsinme**
good day @geoffw0
I want to make changes to this request.
to qualify for the reward program, thank you.

### CODE REVIEW

**Comment by JarLob** [comment]
Thanks!
Few more FP patterns observed:
1.  [An odd, but common usage of an exception for getting the message from `what` function](https://lgtm.com/projects/g/OpenKJ/OpenKJ/snapshot/8611315ea458035e0edd08efe2a55d4497bced5a/files/src/3rdparty/spdlog/include/spdlog/fmt/bundled/format-inl.h?sort=name&dir=ASC&mode=heatmap#L2545). Could it be excluded?
2. Storage of an exception in a [wrapper exception](https://lgtm.com/projects/g/tomkwok/mediasoup-cpp/snapshot/6052b6920bcd24a58da7fc48a0a2322128ef0e22/files/deps/json/single_include/nlohmann/json.hpp?sort=name&dir=ASC&mode=heatmap#L1879) and [sometimes in a class that is not derived from exception](https://lgtm.com/projects/g/tomkwok/mediasoup-cpp/snapshot/6052b6920bcd24a58da7fc48a0a2322128ef0e22/files/deps/json/single_include/nlohmann/json.hpp?sort=name&dir=ASC&mode=heatmap#L1879).

**Comment by ihsinme** [comment]
Thanks.
I will definitely try to fix everything.

**Comment by ihsinme** [comment]
@JarLob 
even if this is a strong enough restriction.
but we will get rid of all false positives.

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql [hunk 1]
```diff
@@ -38,13 +38,6 @@ where
...
   ) and
-  not fc.isInMacroExpansion() and
-  not exists(ThrowExpr texp | fc.getEnclosingStmt() = texp.getEnclosingStmt()) and
-  not exists(FunctionCall fctmp | fctmp.getAnArgument() = fc) and
-  not fc instanceof ConstructorDirectInit and
-  not fc.getEnclosingStmt() instanceof DeclStmt and
-  not fc instanceof ConstructorDelegationInit and
-  not fc.getParent() instanceof Initializer and
-  not fc.getParent() instanceof AllocationExpr and
+  fc instanceof ExprInVoidContext and
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:41]: I think you still want to exclude the case where `fc` is in a macro expansion, right?

Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:41]: thanks

**Review by geoffw0 — APPROVED** [review]
LGTM.

We lose 1 good result (in `godotengine/godot`), where a `std::exception` is created, converted to `int` and returned - which does not seem very sensible.  All the other [results removed by this change](https://lgtm.com/query/2945672473023009122/) look like they were false positives.  I think that's a good tradeoff!

:+1:

**Comment by JarLob** [comment]
Let me re-run it on LGTM before merging.

**Comment by JarLob** [comment]
Results looks good. @geoffw0 @MathiasVP Do you think it is worth to change the message to something like "Object creation of exception type on stack. Did you forget the `throw` keyword?" Anyway I'm ok with the old message too.

**Comment by MathiasVP** [comment]
> Object creation of exception type on stack. Did you forget the throw keyword?

I definitely prefer "Object creation of exception type on stack. Did you forget the throw keyword?" 👍

**Comment by ihsinme** [comment]
@MathiasVP 
I made changes. 
can you run a check?

**Comment by JarLob** [comment]
> I made changes. can you run a check?

Done.

**Comment by ihsinme** [comment]
maybe something else is needed from me for this PR to be accepted?

**Comment by MathiasVP** [comment]
This all looks good to me. I'd be happy to merge this as soon as @JarLob is happy with it.

**Review by JarLob — APPROVED** [review]

**Comment by JarLob** [comment]
LGTM

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**